### PR TITLE
feat(restore): Optimizing restore for incremental backup

### PR DESCRIPTION
**Pull request by mynktl**
**Why is this PR required? What issue does it fix?**:
This PR optimizes restore flow for incremental backups. 

**What this PR does?**:
This PR adds the capability in the plugin to restore all the backups,
including the base and incremental backups, required for the given restore.
With this change, Users don't need to perform multiple restores.
A single restore, for the given backup, will restore all the backups,
from base backup to required backup.

**Does this PR require any upgrade changes?**:
No

**Any additional information for your reviewer?**:
Missing items:
- Updating README.md

**Checklist:**
- [ ] Fixes #<issue number>
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [x] Has the change log section been updated?
- [x] Commit has unit tests
- [x] Commit has integration tests
- [ ] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [ ] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

### CODE REVIEW

File: pkg/cstor/pv_operation.go [hunk 1]
```diff
@@ -158,3 +198,35 @@ func isCSIPv(pv v1.PersistentVolume) bool {
 	}
 	return false
 }
+
+// getSnapListFromCloud returns list of all the snapshots(base and incremental) associated with given volume from cloud
+func (p *Plugin) getSnapListFromCloud(vol *Volume) ([]string, error) {
```
Comment (pawanpraka1) [inline pkg/cstor/pv_operation.go:203]: let us move this function to  pkg/clouduploader and pass bkpname and tag (snapshottag), which will return the snapshot list.

File: pkg/cstor/pv_operation.go [hunk 2]
```diff
@@ -58,7 +59,53 @@ func (p *Plugin) updateVolCASInfo(data []byte, volumeID string) error {
...
+		snapshotList = []string{targetBackupName}
+	}
+
+	// snapshots are created using timestamp, we need to sort it in ascending order
+	sort.Strings(snapshotList)
+
+	for _, snap := range snapshotList {
+		p.Log.Infof("Restoring snapshot=%s", snap)
+
+		vol.backupName = snap
```
Comment (vishnuitta) [inline pkg/cstor/pv_operation.go:91]: this is overriding 'vol' which is input to this method. That might be not intended behavior.
can we make a copy of 'vol' before changing it?

File: tests/velero/restore.go [hunk 3]
```diff
@@ -78,35 +78,32 @@ func (c *ClientSet) GetScheduledBackups(schedule string) ([]string, error) {
...
-func (c *ClientSet) CreateRestoreFromSchedule(ns, targetedNs, schedule string, n int) (v1.RestorePhase, error) {
-	var status v1.RestorePhase
-	var err error
+// CreateRestore create restore from given backup/schedule for ns Namespace to targetedNs
+// Arguments:
+// - ns : source namespace or namespace from backup
+// - targetedNs : targeted namespace for namespace 'ns'
+// - backup : name of the backup, from which restore will happen
+// - schedule : name of schedule, from which restore should happen. If mentioned, backup should be empty
+func (c *ClientSet) CreateRestore(ns, targetedNs, backup, schedule string) (v1.RestorePhase, error) {
```
Comment (vishnuitta) [inline tests/velero/restore.go:87]: I don't see the change in caller.. will that be in diff repo?

File: pkg/cstor/pv_operation.go (hunk 2 above)
Comment (mynktl) [reply pkg/cstor/pv_operation.go:91]: `backupName` act as argument for restore. Caller uses `vol` pointer for update on status. I'll change this function to take backupName as arg and update vol.backupName for restore.

File: tests/velero/restore.go (hunk 3 above)
Comment (mynktl) [reply tests/velero/restore.go:87]: It is in same PR. relevant test case `Restore from scheduled backup`

File: pkg/cstor/pv_operation.go [hunk 4]
```diff
@@ -58,7 +59,51 @@ func (p *Plugin) updateVolCASInfo(data []byte, volumeID string) error {
...
+		if err != nil {
+			return err
+		}
+	} else {
+		// We are restoring only given backup
+		snapshotList = []string{targetBackupName}
+	}
+
+	// snapshots are created using timestamp, we need to sort it in ascending order
+	sort.Strings(snapshotList)
```
Comment (pawanpraka1) [inline pkg/cstor/pv_operation.go:84]: put a check to make sure snapshotList in not empty

File: tests/velero/restore.go [hunk 5]
```diff
@@ -120,6 +117,7 @@ func (c *ClientSet) CreateRestore(ns, targetedNs, backup string) (v1.RestorePhas
 			IncludedNamespaces: []string{ns},
 			RestorePVs:         &snapVolume,
 			BackupName:         backup,
+			ScheduleName:       schedule,
```
Comment (pawanpraka1) [inline tests/velero/restore.go:120]: do we need this?

File: script/volumesnapshotlocation.yaml [hunk 6]
```diff
@@ -26,3 +26,4 @@ spec:
     region: minio
     s3Url: MINIO_ENDPOINT
     s3ForcePathStyle: "true"
+    restoreAllIncrementalSnapshots: "true"
```
Comment (pawanpraka1) [inline script/volumesnapshotlocation.yaml:29]: fullRestore?

File: pkg/cstor/pv_operation.go (hunk 4 above)
Comment (pawanpraka1) [reply pkg/cstor/pv_operation.go:84]: better to check if targetBackupName exist in the snaplist.

File: script/volumesnapshotlocation.yaml (hunk 6 above)
Comment (mynktl) [reply script/volumesnapshotlocation.yaml:29]: fullRestore may confuse with restore of full backup.

File: tests/velero/restore.go (hunk 5 above)
Comment (mynktl) [reply tests/velero/restore.go:120]: yes, for test case [**_Restore from scheduled backup to different Namespace_**](https://github.com/openebs/velero-plugin/pull/99/files/7553cc6bb4ad69ea4d67c17652d9689dd97f9005#diff-60de5209d5a0ca5555d0bf45dafe12cf0f7717e517339654e44cb55f4199e9baR166)

File: pkg/cstor/pv_operation.go (hunk 4 above)
Comment (mynktl) [reply pkg/cstor/pv_operation.go:92]: done

**Comment by mynktl** [comment]
Hi @pawanpraka1,

I've incorporated the review comment. PTAL. 

Thanks!

File: pkg/cstor/pv_operation.go [hunk 7]
```diff
@@ -58,7 +59,59 @@ func (p *Plugin) updateVolCASInfo(data []byte, volumeID string) error {
...
+		snapshotList, err = p.cl.GetSnapListFromCloud(vol.snapshotTag, p.getScheduleName(targetBackupName))
+		if err != nil {
+			return err
+		}
+	} else {
+		// We are restoring only given backup
+		snapshotList = []string{targetBackupName}
+	}
+
+	if len(snapshotList) == 0 {
```
Comment (pawanpraka1) [inline pkg/cstor/pv_operation.go:83]: after below check for ` 	if !contains(snapshotList, targetBackupName) `, we don't need this one.

File: tests/velero/restore.go (hunk 5 above)
Comment (pawanpraka1) [reply tests/velero/restore.go:120]: not able to see the changes where we are using variable ScheduleName in this PR?

**Review by pawanpraka1 — COMMENTED** [review]
looks good. see if we can find better name for restoreAllIncrementalSnapshots in namingishard channel.

File: pkg/cstor/pv_operation.go (hunk 7 above)
Comment (mynktl) [reply pkg/cstor/pv_operation.go]: ok

File: README.md [hunk 8]
```diff
@@ -265,17 +265,24 @@ To restore in different namespace, run the following command:
 velero restore create --from-backup backup_name --restore-volumes=true --namespace-mappings source_ns:destination_ns
 ```
 
-*Note: After restore for remote backup is completed, you need to set target-ip for the volume in pool pod. If restore is from local snapshot then you don't need to update target-ip*
-*Steps to get target-ip*
-1. kubectl get svc -n openebs <PV_NAME> -ojsonpath='{.spec.clusterIP}'
-*Steps to update `target-ip` in pool pod is as follow:*
+Once restore for remote backup is completed, You need to set targetip in relevant replica. Refer [Setting targetip in replica](#setting-targetip-in-replica).
```
Comment (kmova) [inline README.md:268]: Make this bold. **Once restore for remote backup is completed, You need to set targetip in relevant replica. Refer [Setting targetip in replica](#setting-targetip-in-replica).**

File: tests/velero/restore.go (hunk 5 above)
Comment (mynktl) [reply tests/velero/restore.go:120]: yes. https://github.com/openebs/velero-plugin/blob/7553cc6bb4ad69ea4d67c17652d9689dd97f9005/tests/sanity/backup_test.go#L166

**Review by vishnuitta — APPROVED** [review]
changes are good..
one qn is: do we need to handle doing restores of failed/partially failed backups?

**Comment by mynktl** [comment]
> changes are good..
> one qn is: do we need to handle doing restores of failed/partially failed backups?

good qn. We need to check for snapshot file existence. `GetSnapListFromCloud` check file using prefix only. For cstor, we are storing snapshot and the PVC both. So in case of failed backup, `GetSnapListFromCloud` will return the failed backup also. I've added check to verify if snapshot file is present or not.

**Review by vishnuitta — APPROVED** [review]
changes are good

**Review by kmova — APPROVED** [review]

**Pull request merged**